common/dbus/platform/asio_libdbus/dbus_client.cpp

```diff
@@ -0,0 +1,307 @@
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include <common/error.hpp>
+#include <common/expected.hpp>
+#include <common/events.hpp>
+#include <common/log.hpp>
+
+#include <boost/asio.hpp>
+#include <dbus/dbus.h>
+
+namespace mender {
+namespace common {
+namespace dbus {
+
+namespace asio = boost::asio;
+
+namespace error = mender::common::error;
+namespace expected = mender::common::expected;
+namespace events = mender::common::events;
+namespace log = mender::common::log;
+
+using namespace std;
+
+// The code below integrates ASIO and libdbus. Or, more precisely, it uses
+// asio::io_context as the main/event loop for libdbus.
+//
+// HandleDispatch() makes sure message dispatch is done. The *Watch() functions
+// allow libdbus to set up and cancel watching of its connection's file
+// descriptor(s). The *Timeout() functions do the same just for
+// timeouts. HandleReply() is a C function we use to extract the DBus reply and
+// pass it to a handler given to DBusClient::CallMethod().
+// (see the individual functions below for details)
+
+// friends can't be static (see class DBusClient)
+void HandleDispatch(DBusConnection *conn, DBusDispatchStatus status, void *data);
+dbus_bool_t AddDBusWatch(DBusWatch *w, void *data);
+static void RemoveDBusWatch(DBusWatch *w, void *data);
+static void ToggleDBusWatch(DBusWatch *w, void *data);
+dbus_bool_t AddDBusTimeout(DBusTimeout *t, void *data);
+static void RemoveDBusTimeout(DBusTimeout *t, void *data);
+static void ToggleDBusTimeout(DBusTimeout *t, void *data);
+static void HandleReply(DBusPendingCall *pending, void *data);
+
+error::Error DBusClient::InitializeConnection() {
+	DBusError dbus_error;
+	dbus_error_init(&dbus_error);
+	dbus_conn_.reset(dbus_bus_get(DBUS_BUS_SYSTEM, &dbus_error));
+	if (!dbus_conn_) {
+		return MakeError(
+			ConnectionError,
+			string("Failed to get connection to system bus: ") + dbus_error.message + "["
+				+ dbus_error.name + "]");
+	}
+
+	dbus_connection_set_exit_on_disconnect(dbus_conn_.get(), FALSE);
+	if (!dbus_connection_set_watch_functions(
+			dbus_conn_.get(), AddDBusWatch, RemoveDBusWatch, ToggleDBusWatch, this, NULL)) {
+		dbus_conn_.reset();
+		return MakeError(ConnectionError, "Failed to set watch functions");
+	}
+	if (!dbus_connection_set_timeout_functions(
+			dbus_conn_.get(), AddDBusTimeout, RemoveDBusTimeout, ToggleDBusTimeout, this, NULL)) {
+		dbus_conn_.reset();
+		return MakeError(ConnectionError, "Failed to set timeout functions");
+	}
+	dbus_connection_set_dispatch_status_function(dbus_conn_.get(), HandleDispatch, this, NULL);
+
+	return error::NoError;
+}
+
+void FreeHandlerCopy(void *data) {
+	DBusCallReplyHandler *handler = static_cast<DBusCallReplyHandler *>(data);
+	delete handler;
+}
+
+error::Error DBusClient::CallMethod(
+	const string &destination,
+	const string &path,
+	const string &iface,
+	const string &method,
+	DBusCallReplyHandler handler) {
+	// TODO: with a mutex to make this thread-safe?
```

Comment:
I would perhaps suggest that it is documented as not thread safe instead, and the caller should use a separate instance if you need to utilize several threads. I'm not even sure how the API would work if it could be accessed from several threads simultaneously..

Reply:
:+1: Actually, thread-safety and (true) absence of race conditions was among the things often mentioned by the authors of the other DBus implementations not based on *libdbus*.